core/src/test/scala/org/apache/spark/storage/MemoryStoreSuite.scala

```diff
@@ -526,4 +526,84 @@ class MemoryStoreSuite
       }
     }
   }
+
+  test("[SPARK-24225]: remove should close AutoCloseable object") {
+
+    val (store, _) = makeMemoryStore(12000)
+
+    val id = BroadcastBlockId(0)
+    val tracker = new CloseTracker()
+    store.putIteratorAsValues(id, Iterator(tracker), ClassTag.Any)
+    assert(store.contains(id))
+    store.remove(id)
+    assert(tracker.getClosed())
+  }
+
+  test("[SPARK-24225]: remove should close AutoCloseable objects even if they throw exceptions") {
+
+    val (store, _) = makeMemoryStore(12000)
+
+    val id = BroadcastBlockId(0)
+    val tracker = new CloseTracker(true)
+    store.putIteratorAsValues(id, Iterator(tracker), ClassTag.Any)
+    assert(store.contains(id))
+    store.remove(id)
+    assert(tracker.getClosed())
+  }
+
+  test("[SPARK-24225]: clear should close AutoCloseable objects") {
+
+    val (store, _) = makeMemoryStore(12000)
+
+    val id = BroadcastBlockId(0)
+    val tracker = new CloseTracker
+    store.putIteratorAsValues(id, Iterator(tracker), ClassTag.Any)
+    assert(store.contains(id))
+    store.clear()
+    assert(tracker.getClosed())
+  }
+
+  test("[SPARK-24225]: clear should close all AutoCloseable objects put together in an iterator") {
```

Comment:
Can you check if you have multiple autocloseable objects in an iterator, and only one of them is removed, the rests of the objects will not be closed?

Reply:
So if I understand the API correctly, there is no way to remove a single item that was put as part of a call to `putIterator` - because operations are conducted by `blockId` you would only be able to remove the whole group of entries, not a single part of an iterator.